feat: fastapi node health status endpoint BLD-1317

giddyphysicist opened the pull request:
## Summary
<!-- Describe the changes in this PR. -->
Adds ability to import a health status endpoint to an Aineko FastAPI app. 

Endpoint located at route `/health` will return a 200 status code if the FastAPI server is active. 

How to implement -- add the following lines to the FastAPI main app file:
```python
from aineko.extras.fastapi import health_router

app.include_router(health_router)
```
where `app` is the user instance of an Aineko FastAPI app.

## Motivation
<!-- Describe the motivation for this change. -->
Previously, there was not a general method for checking the status for an Aineko pipeline FastAPI node. 

This shows that the FastAPI Aineko node and the pipeline are active. 

## Author Checklist

- [x] Changes are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Documentation is updated if necessary.
- [x] Status checks pass.
- [ ] If the version number is changed, it is updated in `pyproject.toml` , `aineko/__init__.py`, `aineko/templates/first_aineko_pipeline/{{cookiecutter.project_slug}}/pyproject.toml`.


## Reviewer Checklist

- [x] Title is accurate and formatted appropriately.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Docs are updated if necessary.
- [x] Code is pythonic and consistent with the rest of the codebase.
- [x] The code is consistent with the [style guide](https://google.github.io/styleguide/pyguide.html).

giddyphysicist committed:
feat: fastapi health router endpoint

giddyphysicist committed:
test: added health endpoint to integration test

giddyphysicist committed:
docs: added health check endpoint to docs

giddyphysicist committed:
chore: docs update and endpoint status

added explicit endpoint status code in argument for endpoint, and added async def.

giddyphysicist committed:
Merge branch 'develop' into feat/api-health

giddyphysicist committed:
chore: add middleware to spelling list

mdigi14 commented on aineko/extras/fastapi/health.py:
```diff
@@ -0,0 +1,33 @@
+
+
+@router.get("/health", status_code=200)
+async def read_health() -> dict:
```
I don't think this needs to be `async`. We're not using `await` so it doesn't add anything.

Barabazs replied:
I think FastAPI recommends to always use `async def` unless you need to blocking operations.
 
https://fastapi.tiangolo.com/async/#path-operation-functions

mdigi14 replied:
@Barabazs Thanks! good to know

mdigi14 commented on aineko/extras/fastapi/health.py:
````diff
@@ -0,0 +1,33 @@
+    a GET request query to `http://localhost:8000/health`.
+
+    Example usage in FastAPI app:
+        ```python hl_lines="2 5"
````
Is this `hl_lines` syntax related to MKdocs? Im curious how I can preview changes like this on PRs

Barabazs replied:
You can run `make view-docs` locally and then navigate to the page documents this file.
In this case it's to highlight line 2 and 5.

giddyphysicist replied:
oh, nice. i didn't know we had a make command for that. I had been using `poetry run mkdocs serve` locally

mdigi14 commented on aineko/extras/fastapi/health.py:
```diff
@@ -0,0 +1,33 @@
+
+from fastapi import APIRouter
+
+router = APIRouter()
```
We can probably just call this `health_router` and then use 

`from aineko.extras.fastapi.health import health_router `

giddyphysicist replied:
updated here: [42dc84c](https://github.com/aineko-dev/aineko/pull/142/commits/42dc84c8995f32e0b2a788e40c2903ee5a29eb10)

Barabazs commented on docs/extras/fastapi.md:
```diff
@@ -11,3 +11,15 @@ The FastAPI extra node can be used by adding the following to `pyproject.toml`
+
+## Authentication Considerations
+
+By default, the Aineko health endpoint does not contain authentication. 
```
```suggestion
By default, the `/health` endpoint does not contain authentication. 
```
or just
```suggestion
By default, the `health` endpoint does not contain authentication. 
```

Imo we should drop the Aineko here.

giddyphysicist replied:
updated here: [42dc84c](https://github.com/aineko-dev/aineko/pull/142/commits/42dc84c8995f32e0b2a788e40c2903ee5a29eb10)

Barabazs commented on docs/extras/fastapi.md:
```diff
@@ -11,3 +11,15 @@ The FastAPI extra node can be used by adding the following to `pyproject.toml`
+
+By default, the Aineko health endpoint does not contain authentication. 
+
+One method to add authentication is to use security middleware at the app level for FastAPI. This will handle adding authentication to the health endpoint without having to add custom dependencies.
```
Maybe we should link the FastAPI docs here: https://fastapi.tiangolo.com/advanced/security/

giddyphysicist replied:
updated here: [42dc84c](https://github.com/aineko-dev/aineko/pull/142/commits/42dc84c8995f32e0b2a788e40c2903ee5a29eb10)

Barabazs commented on docs/extras/fastapi.md:
I think we should include a short paragraph about the reason to use this health endpoint.

giddyphysicist replied:
update here: [42dc84c](https://github.com/aineko-dev/aineko/pull/142/commits/42dc84c8995f32e0b2a788e40c2903ee5a29eb10)

giddyphysicist committed:
chore: address PR feedback

- changed name of router and import
- added summary of use case for endpoint in docs
- removed 'aineko' in docs
- added link to fastapi security docs in docs

programmablemike reviewed (approved):
👍 LGTM. Thanks for submitting it!
